File: src/main/java/net/orcinus/galosphere/client/particles/IndicatorParticle.java
package net.orcinus.galosphere.client.particles;

import com.mojang.blaze3d.vertex.VertexConsumer;
import com.mojang.math.Axis;
import net.minecraft.client.Camera;
import net.minecraft.client.multiplayer.ClientLevel;
import net.minecraft.client.particle.Particle;
import net.minecraft.client.particle.ParticleProvider;
import net.minecraft.client.particle.ParticleRenderType;
import net.minecraft.client.particle.SpriteSet;
import net.minecraft.client.particle.TextureSheetParticle;
import net.minecraft.core.BlockPos;
import net.minecraft.core.particles.SimpleParticleType;
import net.minecraft.util.Mth;
import net.minecraft.world.phys.Vec3;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.orcinus.galosphere.blocks.MonstrometerBlock;
import org.joml.Vector3f;

@OnlyIn(Dist.CLIENT)
public class IndicatorParticle extends TextureSheetParticle {

    private final SpriteSet sprites;

    private static final Vector3f SAFE_COLOR = Vec3.fromRGB24(0xFFB219).toVector3f();
    private static final Vector3f UNSAFE_COLOR = Vec3.fromRGB24(0x93B9FF).toVector3f();

    public IndicatorParticle(ClientLevel world, double x, double y, double z, double pQuadSizeMulitiplier, SpriteSet sprites) {
        super(world, x, y, z);

        alpha = 1;
        quadSize = 0;
        lifetime = 48;

        rCol = getColor().x();
        gCol = getColor().y();
        bCol = getColor().z();

        setSpriteFromAge(this.sprites = sprites);
    }

    private Vector3f getColor() {
        return MonstrometerBlock.isUnsafe(level, BlockPos.containing(x, y, z)) ? UNSAFE_COLOR : SAFE_COLOR;
    }

    @Override
    public void tick() {
        xo = x;
        yo = y;
        zo = z;

        quadSize = Mth.lerp(0.06F, quadSize, 0.5F);

        rCol = Mth.lerp(0.25F, rCol, getColor().x());
        gCol = Mth.lerp(0.25F, gCol, getColor().y());
        bCol = Mth.lerp(0.25F, bCol, getColor().z());

        if (age++ >= lifetime) {
            remove();
        } else {
            if (age > (lifetime / 2)) {
                alpha -= 0.04F;
            }
        }

        setSpriteFromAge(sprites);
    }

    @Override
    public ParticleRenderType getRenderType() {
        return ParticleRenderType.PARTICLE_SHEET_TRANSLUCENT;
    }

    @Override
    public void render(VertexConsumer consumer, Camera camera, float delta) {
        this.renderRotatedQuad(consumer, camera, Axis.XP.rotation(Mth.PI / 2), delta);
        this.renderRotatedQuad(consumer, camera, Axis.XN.rotation(Mth.PI / 2), delta);
    }

    @Override
    protected int getLightColor(float tint) {
        return Math.max(50, super.getLightColor(tint));
    }

    @OnlyIn(Dist.CLIENT)
    public static class Provider implements ParticleProvider<SimpleParticleType> {
        private final SpriteSet sprites;

        public Provider(SpriteSet sprites) {
            this.sprites = sprites;
        }

        public Particle createParticle(SimpleParticleType pType, ClientLevel pLevel, double pX, double pY, double pZ, double speed, double pYSpeed, double pZSpeed) {
            return new IndicatorParticle(pLevel, pX, pY, pZ, speed, sprites);
        }
    }
}

File: src/main/java/net/orcinus/galosphere/client/particles/IndicatorParticle.kt
package net.orcinus.galosphere.client.particles

import com.mojang.blaze3d.vertex.VertexConsumer
import com.mojang.math.Axis
import net.minecraft.client.Camera
import net.minecraft.client.multiplayer.ClientLevel
import net.minecraft.client.particle.*
import net.minecraft.core.BlockPos
import net.minecraft.core.particles.SimpleParticleType
import net.minecraft.util.Mth
import net.minecraft.world.phys.Vec3
import net.neoforged.api.distmarker.Dist
import net.neoforged.api.distmarker.OnlyIn
import net.orcinus.galosphere.blocks.MonstrometerBlock
import org.joml.Vector3f
import kotlin.math.max

@OnlyIn(Dist.CLIENT)
class IndicatorParticle(
	world: ClientLevel,
	x: Double,
	y: Double,
	z: Double,
	pQuadSizeMulitiplier: Double,
	sprites: SpriteSet
) : TextureSheetParticle(world, x, y, z)
{
	private val sprites: SpriteSet

	init
	{
		alpha = 1f
		quadSize = 0f
		lifetime = 48

		rCol = this.color.x()
		gCol = this.color.y()
		bCol = this.color.z()

		setSpriteFromAge(sprites.also { this.sprites = it })
	}

	private val color
		get() = if (MonstrometerBlock.isUnsafe(level, BlockPos.containing(x, y, z)))
			UNSAFE_COLOR
		else
			SAFE_COLOR

	override fun tick()
	{
		xo = x
		yo = y
		zo = z

		quadSize = Mth.lerp(0.06f, quadSize, 0.5f)

		rCol = Mth.lerp(0.25f, rCol, this.color.x())
		gCol = Mth.lerp(0.25f, gCol, this.color.y())
		bCol = Mth.lerp(0.25f, bCol, this.color.z())

		if (age++ >= lifetime)
		{
			remove()
		}
		else
		{
			if (age > (lifetime / 2))
			{
				alpha -= 0.04f
			}
		}

		setSpriteFromAge(sprites)
	}

	override fun getRenderType(): ParticleRenderType
	{
		return ParticleRenderType.PARTICLE_SHEET_TRANSLUCENT
	}

	override fun render(consumer: VertexConsumer, camera: Camera, delta: Float)
	{
		this.renderRotatedQuad(consumer, camera, Axis.XP.rotation(Mth.PI / 2), delta)
		this.renderRotatedQuad(consumer, camera, Axis.XN.rotation(Mth.PI / 2), delta)
	}

	override fun getLightColor(tint: Float): Int
	{
		return max(50, super.getLightColor(tint))
	}

	@OnlyIn(Dist.CLIENT)
	class Provider(private val sprites: SpriteSet) : ParticleProvider<SimpleParticleType>
	{
		override fun createParticle(
			pType: SimpleParticleType,
			pLevel: ClientLevel,
			pX: Double,
			pY: Double,
			pZ: Double,
			speed: Double,
			pYSpeed: Double,
			pZSpeed: Double
		): Particle
		{
			return IndicatorParticle(pLevel, pX, pY, pZ, speed, sprites)
		}
	}

	companion object
	{
		private val SAFE_COLOR = Vec3.fromRGB24(0xFFB219).toVector3f()
		private val UNSAFE_COLOR = Vec3.fromRGB24(0x93B9FF).toVector3f()
	}
}